### PR TITLE
new dpu: E-SparqlEndpoint

### PR DESCRIPTION
**Pull request by skodape**
Pull request for yet not finished DPU for RDF extraction from SPARQL endpoint. Connects to  #178

What needs to be done:
1) SPARQL validation same as in SPARQL update/construct
2) As this DPU should replace SPARQL Endpoint Extractor from Charles University repository, we need to check against used triple stores and make sure that this DPU works fine - we should test against Sesame virtuoso fuseki and @jindrichmynarz please specify whether we need also to extract data from parliament store
3) sk-localization
4) make sure that latest parent project is used

@jakubklimek Should be asked to provide instances of endpoints to test on.

### CODE REVIEW

File: e-sparqlEndpoint/src/main/java/cz/cuni/mff/xrg/uv/extractor/sparqlendpoint/RemoteRdfDataUnit.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    /**
+     * From: https://groups.google.com/forum/#!topic/sesame-users/5UYsqct3y-w
+     *
+     * Virtuoso does not implement the full Sesame Server protocol (which is an
+     * extension of the SPARQL protocol), so you should not be using HTTPRepository to
+     * connect to it. Instead, use SPARQLRepository, which is specifically designed to
+     * connect to third-party SPARQL endpoints:
+     *
+     */
+    private final SPARQLRepository repository;
```
Comment (mkl-eea-sk) [inline e-sparqlEndpoint/src/main/java/cz/cuni/mff/xrg/uv/extractor/sparqlendpoint/RemoteRdfDataUnit.java]: congratz!!!!

**Comment by skrchnavy** [comment]
Connected to #178

File: e-sparqlEndpoint/pom.xml [hunk 2]
```diff
@@ -0,0 +1,120 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+	<parent>
+		<groupId>eu.unifiedviews</groupId>
+		<artifactId>uv-pom-dpu</artifactId>
+		<version>2.0.1</version>
+		<relativePath />
+	</parent>
+	<groupId>cz.cuni.mff.xrg.uv</groupId>
```
Comment (skrchnavy) [inline e-sparqlEndpoint/pom.xml]: please move to `eu.unifiedviews.plugins` group ID.

File: e-sparqlEndpoint/pom.xml [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	<artifactId>e-sparqlEndpoint</artifactId>
+	<version>1.0.1</version>
+	<packaging>bundle</packaging>
+	<name>E-SparqlEndpoint</name>
+	<description>Extract RDF data from remote SPARQL.</description>
+
+	<properties>
+		<dpu.package>cz.cuni.mff.xrg.uv.extractor.sparqlendpoint</dpu.package>
```
Comment (skrchnavy) [inline e-sparqlEndpoint/pom.xml]: Refactoring to `eu.unifiedviews.plugins.extractor....` is expected

File: e-sparqlEndpoint/pom.xml [hunk 4]
```diff
@@ -0,0 +1,120 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+	<parent>
+		<groupId>eu.unifiedviews</groupId>
+		<artifactId>uv-pom-dpu</artifactId>
+		<version>2.0.1</version>
```
Comment (skrchnavy) [inline e-sparqlEndpoint/pom.xml]: @tomas-knap is it expected to be compatible with UV 2.0 or UV 2.1 ?

File: pom.xml [hunk 5]
```diff
@@ -10,6 +10,7 @@
     <!-- list of core modules -->
     <modules>
         <!-- EXTRACTORS -->
+		<module>e-sparqlEndpoint</module>
```
Comment (skrchnavy) [inline pom.xml]: Please use spaces in formating of xml

File: e-sparqlEndpoint/pom.xml [hunk 6]
```diff
@@ -0,0 +1,120 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
```
Comment (skrchnavy) [inline e-sparqlEndpoint/pom.xml]: Spaces are recommended for pom.xml

File: e-sparqlEndpoint/pom.xml [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+			<artifactId>opencsv</artifactId>
+			<version>3.3</version>
+		</dependency>
+		<dependency> <!-- Required by opencsv. -->
+			<groupId>org.apache.commons</groupId>
+			<artifactId>commons-lang3</artifactId>
+			<version>3.3.2</version>
+		</dependency>
+	</dependencies>
+
```
Comment (skrchnavy) [inline e-sparqlEndpoint/pom.xml]: missing 
```
    <repositories>
        <repository>
            <id>maven.eea.sk</id>
            <name>maven.eea.sk-public</name>
            <url>http://maven.eea.sk/artifactory/public/</url>
        </repository>
    </repositories>
```
to be able to download parent pom.

File: e-sparqlEndpoint/pom.xml [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+		</dependency>
+		<dependency> <!-- Required by sesame-queryresultio-text. -->
+			<groupId>com.opencsv</groupId>
+			<artifactId>opencsv</artifactId>
+			<version>3.3</version>
+		</dependency>
+		<dependency> <!-- Required by opencsv. -->
+			<groupId>org.apache.commons</groupId>
+			<artifactId>commons-lang3</artifactId>
+			<version>3.3.2</version>
```
Comment (skrchnavy) [inline e-sparqlEndpoint/pom.xml]: version is reused from parent pom dependency  management so this is duplicated.

**Comment by jindrichmynarz** [comment]
@skodape: In order to support COMSODE pilot with the Slovak Environmental Agency we need e-sparqlEndpoint to support Parliament RDF store. In order to do that, not much needs be done. The requests the DPU sends must be provided with the correct MIME type as described in mff-uk/DPUs#104. I presume that doing that would not break support for the other RDF stores, but we should test that assumption.

**Comment by skodape** [comment]
It might be easy .. we use SparqlRepository (as Michal pointed out) as out of the box solution. There is method called ```setAdditionalHttpHeaders``` that may suite our needs. 

But still before usage of this DPU we have to solve aforementioned issues as well. 

It's up to @tomas-knap to choose how to approach those issues and negotiate resource allocation, based on task priorities.

**Comment by tomas-knap** [comment]
@skodape Please discuss with Martin, whether/when you can work on points 1) 2) - at least please test virtuoso, sesame, fuseki (for testing Fuseki, please ensure inputs from @jakubklimek as we discussed). When you discuss with Martin, please write here when it can be done. 

We will then test parliament store (@jindrichmynarz) and also ensure 3) 4), so that we can release the DPU soon - this is rather crucial DPU, which is not in UV 2.X since March, so we should finish that soon. 

Thanks!

**Comment by tomas-knap** [comment]
As a minimalistic solution (if @skodape cannot work on that) is to check that it can extract data from virtuoso repository (+l-parliament). Later on, we can test/adjust the DPU to work also with fuseki/sesame.

**Comment by tomas-knap** [comment]
Summary of TODOs: 
1) SPARQL validation same as in SPARQL update/construct 
2) The logic of the DPU is taken from https://github.com/mff-uk/DPUs/blob/master/dpu/e-sparqlEndpoint/), which works NOW correctly for CUNI against virtuoso, sesame, fuseki (but with the DPU taken from mff-uk/DPUs). Thus, it has to be only tested against parliament store (@jindrichmynarz can do that, when other points are done). No need to change any logic inside. 
3) sk-localization
4) make sure that latest parent project is used
5) https://github.com/UnifiedViews/Plugins/issues/241
6) Take into account comments above and  test the DPU at least against Virtuoso and parliament store (@jindrichmynarz can test against parliament store). 

Points 1) - 4) are required by https://github.com/UnifiedViews/Plugins/issues/178
Point 5) is required by https://github.com/UnifiedViews/Plugins/issues/241

**Comment by skrchnavy** [comment]
*3) make sure that latest parent project is used*
- Done

**Comment by skodape** [comment]
* DPU is ready for localization (properties file is presented)
* Added support for SPARQL validation as in SPARQL construct/update
* DPU tested with simple query ```CONSTRUCT { ?s ?p ?o } WHERE { ?s ?p ?o }``` against following repositories on small data (1,6K triples):
 * Virtuoso 
 * Sesame (Sesame workbench) version 2.8.4
 * Fuseki version 2.0.0
 * Parliament version 2.7.9 (with default graph)

**Pull request merged**